Draft PR: Update to Vert.x 4

ebullient opened the pull request:
This is a collaborative branch: 
* Base is the microprofile-4.0 branch

Commits have been broken into topic areas. Tests are "unfixed" (things will break).

Making sure everything compiles: 
```
> mvn install -DskipTests=true -DskipIT=true --fail-at-end
```

Individual issues will be opened against this PR with instructions on how to recreate

jmartisk committed:
MP Metrics 3.0

jmartisk committed:
MP Metrics 3.0

Ladicek committed:
Update to MicroProfile Fault Tolerance 3.0-RC2 and SmallRye Fault Tolerance 5.0.0-RC1

Ladicek committed:
Update to MicroProfile Fault Tolerance 3.0-RC2 and SmallRye Fault Tolerance 5.0.0-RC1

radcortez committed:
MP Config 2.0 Support.

radcortez committed:
MP Config 2.0 Support.

radcortez committed:
MP JWT 1.2 Support.

radcortez committed:
MP JWT 1.2 Support.

phillip-kruger committed:
Update SmallRye GraphQL to 1.1.0-RC1
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

jmartisk committed:
Exclude two tests that fail on the CI

xstefank committed:
MP Health 3.0

Ladicek committed:
Update to MicroProfile Fault Tolerance 3.0 and SmallRye Fault Tolerance 5.0.0-RC2

jmartisk committed:
Remove support for metrics from CDI producers

Ladicek committed:
Revert "Exclude two tests that fail on the CI"

This reverts commit d3ee21de0107f15277e9eeb663e97b574ef26dbf.

jmartisk committed:
SmallRye Metrics 3.0.0

phillip-kruger committed:
OpenAPI upgrade to 2.1.1 (MP4)
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

radcortez committed:
MP Rest Client 2.0 Support

radcortez committed:
MP Rest Client 2.0 Support

radcortez committed:
MP OpenTracing 2.0 Support.

radcortez committed:
MP OpenTracing 2.0 Support.

xstefank committed:
Update SR Health to 3.0.0

Ladicek committed:
Update to SmallRye Fault Tolerance 5.0.0

phillip-kruger committed:
Upgrade GraphQL to 1.1.0 (MP4)
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

radcortez committed:
Update SR JWT to 3.0.0-RC3.

radcortez committed:
Update SR JWT to 3.0.0-RC3.

cescoffier committed:
Vert.x 4 migration: **/pom.xml

Update Hibernate Reactive version to 1.0.0.Beta4.Vertx4

cescoffier committed:
Vert.x 4: Core / Verticle changes

JSON substitutions moved to undertow

cescoffier committed:
Vert.x 4: Mailer

cescoffier committed:
Vert.x 4: Mutiny and thread pools

cescoffier committed:
Vert.x 4: Redis client

cescoffier committed:
Vert.x 4: OIDC client (temporary)

cescoffier committed:
Vert.x 4: JSON and GraphQL serialization

* Use the proper Base64 encoder
* Substitutions for undertow + JSON
  Undertow explicitly excludes Jackson Core so we need to make sure
  that Vert.x json doesn't trip the native image compilation

cescoffier committed:
Vert.x 4: HttpMethod and .rawMethod() usage

Extend HttpMethod javadoc

cescoffier committed:
Vert.x 4: Http Metrics

cescoffier committed:
Vert.x 4: Http API changes

cescoffier committed:
Vert.x 4: Http/Core event loop

cescoffier committed:
Vert.x 4: Http user authentication

cescoffier committed:
Vert.x 4: Matching of prefixed routes

cescoffier commented:
Thanks @ebullient, great playground!

sberyozkin committed:
Update to Smallrye JWT 3.0.0-RC5

ebullient commented:
Later this afternoon (US time), I'll collapse/regroup commits so far, and will rebase on latest MP4 branch, etc. Will be a force push, so y'all will need a reset to pick up the changes _after having some downtime over the weekend_. :-P

jmartisk committed:
Upgrade to SR Metrics 3.0.1

geoand committed:
Use custom ObjectMapper when quarkus-jackson is present

Fixes: #15031

cescoffier committed:
Disable OIDC integration tests

See https://github.com/quarkusio/quarkus/issues/15084

cescoffier commented on extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/vertx/BufferDeserializer.java:
```diff
@@ -0,0 +1,40 @@
+ */
+package io.quarkus.jackson.runtime.vertx;
+
+import static io.vertx.core.json.impl.JsonUtil.BASE64_DECODER;
```
Because it's in `impl`, I would avoid using it.

cescoffier commented on extensions/jackson/runtime/src/main/java/io/quarkus/jackson/runtime/vertx/BufferSerializer.java:
```diff
@@ -0,0 +1,32 @@
+ */
+package io.quarkus.jackson.runtime.vertx;
+
+import static io.vertx.core.json.impl.JsonUtil.BASE64_ENCODER;
```
Because it's in "impl", I would not use that, but create another instance.

geoand replied:
Done

cescoffier commented:
@geoand It seems that need that the Vert.x SPI is breaking reactive messaging:

```
Error:  io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.testUpdatingConnector  Time elapsed: 5.493 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at io.quarkus.jackson.runtime.vertx.QuarkusJacksonFactory.codec(QuarkusJacksonFactory.java:13)
	at io.vertx.core.json.Json.<clinit>(Json.java:28)
	at io.vertx.core.json.JsonArray.fromJson(JsonArray.java:669)
	at io.vertx.core.json.JsonArray.<init>(JsonArray.java:54)
	at io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.get(ConnectorChangeTest.java:45)
	at io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.lambda$testUpdatingConnector$1(ConnectorChangeTest.java:30)
	at org.awaitility.core.CallableCondition$ConditionEvaluationWrapper.eval(CallableCondition.java:99)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at io.quarkus.jackson.runtime.vertx.QuarkusJacksonJsonCodec.<clinit>(QuarkusJacksonJsonCodec.java:37)
	... 13 more
```

geoand commented:
I'll take a look

geoand committed:
Fix some issues with Vert.x 4 + Jackson integration

geoand commented:
> @geoand It seems that need that the Vert.x SPI is breaking reactive messaging:
> 
> ```
> Error:  io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.testUpdatingConnector  Time elapsed: 5.493 s  <<< ERROR!
> java.lang.ExceptionInInitializerError
> 	at io.quarkus.jackson.runtime.vertx.QuarkusJacksonFactory.codec(QuarkusJacksonFactory.java:13)
> 	at io.vertx.core.json.Json.<clinit>(Json.java:28)
> 	at io.vertx.core.json.JsonArray.fromJson(JsonArray.java:669)
> 	at io.vertx.core.json.JsonArray.<init>(JsonArray.java:54)
> 	at io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.get(ConnectorChangeTest.java:45)
> 	at io.quarkus.smallrye.reactivemessaging.hotreload.ConnectorChangeTest.lambda$testUpdatingConnector$1(ConnectorChangeTest.java:30)
> 	at org.awaitility.core.CallableCondition$ConditionEvaluationWrapper.eval(CallableCondition.java:99)
> 	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
> 	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: java.lang.NullPointerException
> 	at io.quarkus.jackson.runtime.vertx.QuarkusJacksonJsonCodec.<clinit>(QuarkusJacksonJsonCodec.java:37)
> 	... 13 more
> ```

Should be fixed now

cescoffier committed:
Disable OIDC Wiremock tests

radcortez committed:
Update smallrye openapi to 2.1.2

radcortez committed:
Update SmallRye Config to 2.1.0.

cescoffier committed:
Migrate the vertx-graphql extension and its tests to Vert.x 4

- the body handler is now required
- avoid deprecated API
- update tests to follow the ApolloWS protocol which is now enforced (https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md)

radcortez committed:
Fix OpenTracing / RESTEasy native build.

radcortez committed:
Fixed style issues.

sberyozkin committed:
Fix the latest rebase issues

sberyozkin committed:
Update mutiny-vertx.version to 2.1.1 and quarkus-http.version to 4.0.0.Alpha3

sberyozkin committed:
Reenable OIDC tests

sberyozkin committed:
Reenable OIDC ServicePublicKeyTestCase as well

cescoffier committed:
Vert.x 4 - Fix request dispatching when handling mutltipart

cescoffier committed:
Fix swagger-ui route

cescoffier committed:
Fix code template using the Vert.x HttpMethod, which is not an enum anymore

cescoffier committed:
Add missing body handler in the Vert.x GraphQL integration tests

cescoffier committed:
Fix legacy redirect in the health extension

cescoffier committed:
Update Hibernate Reactive to 1.0.0.CR1.Vertx4

cescoffier committed:
Only end the upload when the file have been written to the FS and not when the uploaded file has been read.

cescoffier committed:
Revert "Update Hibernate Reactive to 1.0.0.CR1.Vertx4"

This reverts commit 92c4f62b8184eb57189cce912320ef1c96a909c3.

cescoffier committed:
Disable failing assertion due to https://github.com/quarkusio/quarkus/issues/15593

cescoffier committed:
Extend TCK CI job timeout

cescoffier committed:
Workaround https://github.com/quarkusio/quarkus/issues/15593

ebullient committed:
Micrometer + Vert.x 4.0.3 + #15734

cescoffier committed:
Disable blocking tests in the resteasy reactive TCK - switch to the vertx4 branch

cescoffier committed:
Update vert.x to 4.0.3 and Netty to 4.1.60

cescoffier committed:
Fix Netty substitutions - SSL related.

cescoffier committed:
Implement the newly introduced HttpServerRequestInternal interface in the RequestWrapper and RequestForwarder

cescoffier committed:
Vert.x 4.0.3 requires new substitutions (SSL related)

cescoffier committed:
Switch resteasy reactive to Vert.x 4.0.3

cescoffier committed:
Update to Quarkus HTTP 4.0.0.Alpha4

cescoffier committed:
Update mailer tests because of Vert.x upstream changes

cescoffier committed:
Update Vert.x Mutiny clients to version 2.2.0

cescoffier committed:
Update gRPC version and add compression support in the mutiny bindings

cescoffier committed:
Update netty tcnative version

cescoffier committed:
Add support for the new multipart only option from Vert.x 4.0.3 (which switch back to the Vert.x 3.9.x behavior)

cescoffier committed:
Merge pull request #15708 from cescoffier/vertx-4.0.3

Integration of Vert.x 4.0.3

michalszynkiewicz committed:
Increase netty maxOrder to 3 to make grpc runSize fit chunkSize

cescoffier committed:
Merge pull request #15874 from michalszynkiewicz/v4-grpc-fix

cescoffier committed:
Disable the resteasy reactive TCK - see https://github.com/quarkusio/quarkus/issues/15912

cescoffier committed:
Update Hibernate and Hibernate Reactive to 5.4.29.Final and 1.0.0.CR1.Vertx4

michalszynkiewicz committed:
Change netty.allocator.maxOrder only if grpc is enabled

cescoffier committed:
Replace multi.transform() and multi.collectItems() as they are deprecated

FroMage committed:
Fix DNS failure handling in RR client due to vert.x 4 upgrade

FroMage committed:
Revert "Disable the resteasy reactive TCK - see https://github.com/quarkusio/quarkus/issues/15912"

This reverts commit 4f42de03f89af6d6910d13b30c605ac09267f290.

cescoffier committed:
Merge pull request #15914 from michalszynkiewicz/conditional-max-order

FroMage committed:
RESTEasy Reactive client vert.x 4 fix

We should never throw within onSuccess: nobody's here to catch us if we fall

cescoffier commented:
Closing, it's now backported in the 2.x branch.
